pkg/pillar/cmd/vaultmgr/vaultmgr.go

```diff
@@ -1083,22 +1114,43 @@ func handleVaultKeyFromControllerImpl(ctxArg interface{}, key string,
 		//Log the type of key used for unlocking default vault
 		log.Noticef("%s unlocked using key type %s", defaultVault,
 			etpm.CompareLegacyandSealedKey().String())
-
-		//Mark the default vault as unlocked
-		ctx.defaultVaultUnlocked = true
-
-		//publish vault key to Controller
-		if err := publishVaultKey(ctx, types.DefaultVaultName); err != nil {
-			log.Errorf("Failed to publish Vault Key, %v", err)
+	} else {
+		// We are here if we receive no keys from controller
+		if ctx.defaultVaultUnlocked {
+			// if already unlocked, do nothing
+			return
+		}
+		log.Warnf("Processing empty EncryptedVaultKeyFromController %s", key)
+		// If we had no luck in unlock with sealed key, we did not send keys to the controller,
+		// and we receive no keys from the controller, than we cannot unlock the vault.
+		// Try to remove and re-create default vault now
+		if err := removeDefaultVault(ctx); err != nil {
```

Comment:
Shouldn't you check the file in /persist/ to see that we did not yet send the key?

Reply:
I slightly modified the comment here. If we are inside the branch we receive empty key from zedagent which indicates that we did not receive any keys from controller and check that we did not save file in /persist.

Reply:
Sorry trying understand the implications here. Did we cover all  conditions in which the empty key arrive from controller ? Basically trying to understand if there could be a error condition where we accidentally remove the vault.

Reply:
@zedi-pramodh that is what I'm also trying to determine. It is a bit hard to see since the logic is spread across multiple places.

Reply:
@eriknordmark I am not convinced we have to implement this code. Better safe than sorry. Why can't we just error out or sent alerts without removing the vault if key is empty ?  It can be a support call if it is really an issue and we will at least know what we are doing at that point. Deleting vault scares me !!

Reply:
We've had a user run into this failure and it tooks weeks of developer time to figure out a likely reason for the failure.
And this can happen again if a device is installed and then before it can connect to the controller to upload the encrypted vault key, the device is powered off and e.g., the boot list in the BIOS is changed to remove the USB stick (which results in the PCRs changing).

So at a very minimum we should add very explicit logs here indicating that the vault needs to be manually deleted.
But it would be good to find out a bullet proof way to tell that the encrypted vault key was never successfully sent to the controller and the controller is not handing back an encrypted vault key, and in only that case delete and recreate the vault. We can add other checks such as there being volumes created as additional safety.

Reply:
@zedi-pramodh See above.

Reply:
I see then at least delete vault  part should be removed.  We can have rest of the code in place and log error messages to catch/report this issue. That immediately catches the attention of developers and/or support team. After that manual triage and recovery is the safest option.

Reply:
Based on the new file set by the installer this is safe enough for me.